### PR TITLE
[12.0] mig account_financial_report

### PR DESCRIPTION
**Pull request by eLBati**
Rebuilt branch from 11.0 and applied commits from https://github.com/OCA/account-financial-reporting/pull/509

TODO
 - [x] https://github.com/OCA/account-financial-reporting/pull/509#pullrequestreview-206896535
 - [x] https://github.com/OCA/account-financial-reporting/pull/509#issuecomment-495672367
 - [x] https://github.com/OCA/account-financial-reporting/pull/555
 - [x] https://github.com/OCA/account-financial-reporting/pull/509#issuecomment-491489410
 - [x] https://github.com/OCA/account-financial-reporting/pull/559#issuecomment-497226416 (Fekete)
 - [x] Translations (https://github.com/OCA/account-financial-reporting/pull/559#issuecomment-497230293)

### CODE REVIEW

**Comment by eLBati** [comment]
@feketemihai @mpanarin  see https://github.com/OCA/account-financial-reporting/pull/559#issuecomment-497226416 Could you tell me which email should I set ?

**Comment by pedrobaeza** [comment]
Have you included all the mentioned PRs?

**Comment by eLBati** [comment]
@qgroulard  I was not able to apply 8796d6f1077ed51161e205178b27475d7d186d99 
How did you generate that? Could you propose another PR to my branch please?

**Comment by eLBati** [comment]
@pedrobaeza Merged PRs are incuded.
I put the TODO list in description. Please have a look

**Comment by pedrobaeza** [comment]
OK, thanks for gathering this. Let's push for have it in v12!

**Comment by mpanarin** [comment]
> @feketemihai @mpanarin see [#559 (comment)](https://github.com/OCA/account-financial-reporting/pull/559#issuecomment-497226416) Could you tell me which email should I set ?

@eLBati `mykhailopanarin@gmail.com`

**Comment by eLBati** [comment]
@mpanarin thanks, fixed

**Comment by MiquelRForgeFlow** [comment]
@eLBati Please, remove https://github.com/OCA/account-financial-reporting/pull/554 from your TODO.

**Comment by eLBati** [comment]
@mreficent done, thanks.

I have a problem with `account_financial_report/static/src/scss/report.scss`: I don't understand why it is not applied to PDF reports. Can anyone have a look?

About translations, I updated `account_financial_report.pot` and `it.po`

**Comment by tate11** [comment]
my opinion for PDF reports necessary inherit_id="web.report_assets_common"

**Comment by eLBati** [comment]
@tate11 thanks! Fixed.

This is now ready for me

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM! :+1: 

Functionally tested

File: account_financial_report/view/report_template.xml
Comment (pedrobaeza) [inline account_financial_report/view/report_template.xml]: Why do you need to change this if you have already changed https://github.com/OCA/account-financial-reporting/pull/559/files#diff-1d943c35c14f57134d9ad64496c5f9b5R6?

Comment (eLBati) [reply account_financial_report/view/report_template.xml]: @pedrobaeza I just removed the SCSS from `account_financial_report.internal_layout` because useless (https://github.com/OCA/account-financial-reporting/compare/7130857bce83b818ac91fe01209f473e271a3759..7c8c6ad5406fa9318e5a7be70b346a0c99a96494)

Removing it from `web.assets_common` would make the HTML report style fail. Removing from `web.report_assets_common` would make the PDF  style fail

Comment (pedrobaeza) [reply account_financial_report/view/report_template.xml]: But why don't you simply keep it as in current v11? I don't want to have a background in the screen for "View" option, and that's what is happening with scss compiled file. This is a large discussion and finally I have removed variables in v11, so no need to put scss, but simply css.

Any way, my fear putting this is that this alters other styles in the backend.

Comment (eLBati) [reply account_financial_report/view/report_template.xml]: @pedrobaeza so you propose to move back to CSS removing the use of `$o-view-background-color` and `$o-main-text-color`, right?

Comment (pedrobaeza) [reply account_financial_report/view/report_template.xml]: Yes, and keep the 2 assets links as they were.

Comment (eLBati) [reply account_financial_report/view/report_template.xml]: Thanks, done

**Comment by eLBati** [comment]
@OCA/accounting-maintainers what do you think?

**Comment by pedrobaeza** [comment]
Reviewing this today for merging. Coverage is still decreasing a lot for such a critical module, but it's not your fault. Also I would want it to remove all the SQL mess, but one step at a time...

File: account_financial_report/tests/abstract_test.py
Comment (pedrobaeza) [inline account_financial_report/tests/abstract_test.py:26]: Don't create this account template, as it's now auto-created from prefix.

File: account_financial_report/tests/abstract_test.py
Comment (pedrobaeza) [inline account_financial_report/tests/abstract_test.py:41]: No need of this (see above).

File: account_financial_report/report/abstract_report_xlsx.py
Comment (pedrobaeza) [inline account_financial_report/report/abstract_report_xlsx.py]: Better to use `fields.Date.to_string`

File: account_financial_report/report/abstract_report.py
Comment (pedrobaeza) [inline account_financial_report/report/abstract_report.py]: Use `AsIs(self._table)` for avoiding problems and proper identification. You need at the beginning:
```python
from psycopg2.extensions import AsIs
```

File: account_financial_report/menuitems.xml
Comment (pedrobaeza) [inline account_financial_report/menuitems.xml]: Remove this instead of commenting

File: account_financial_report/menuitems.xml
Comment (pedrobaeza) [inline account_financial_report/menuitems.xml]: What does this mean of hiding PDF reports? Isn't in v12 the reports already gone?

File: account_financial_report/tests/test_general_ledger.py
Comment (pedrobaeza) [inline account_financial_report/tests/test_general_ledger.py]: Use `fields.Date.from_string` in all of this.

File: account_financial_report/wizard/general_ledger_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/general_ledger_wizard.py]: Use `fields.Date.to_string`

File: account_financial_report/wizard/trial_balance_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/trial_balance_wizard.py]: Use `fields.Date.to_string`

File: account_financial_report/report/abstract_report.py
Comment (pedrobaeza) [inline account_financial_report/report/abstract_report.py]: Avoid here SQL injection as done in other change

File: account_financial_report/report/abstract_report_xlsx.py
Comment (pedrobaeza) [inline account_financial_report/report/abstract_report_xlsx.py]: `fields.Date.to_string`

File: account_financial_report/report/open_items.py
Comment (pedrobaeza) [inline account_financial_report/report/open_items.py]: Instead of disabling pylint, use `AsIs` for passing a full expression without warnings

File: account_financial_report/report/trial_balance.py
Comment (pedrobaeza) [inline account_financial_report/report/trial_balance.py:80]: Put a comment in code about why is this a char (due to `parent_path`)

File: account_financial_report/tests/abstract_test.py
Comment (pedrobaeza) [inline account_financial_report/tests/abstract_test.py]: Not sure if a previous error in this method is still there in v12. See https://github.com/OCA/bank-payment/commit/9c5a98ffbe72531098b7c10c676b2c616284fd9b for details.

You can test the problem when having other chart template installed in the database doing the tests.

Any way, previous code is missing explicitly in the creation dictionary `'sale_tax_id' : False` and `'purchase_tax_id': False`, which avoids the problem.

File: account_financial_report/tests/test_general_ledger.py
Comment (pedrobaeza) [inline account_financial_report/tests/test_general_ledger.py]: `fields.Date.from_string('2014-05-05')`
or even better:
`date(2014, 5, 5)`

File: account_financial_report/wizard/aged_partner_balance_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/aged_partner_balance_wizard.py]: This part of the domain is not correct at all: you can have subsidiaries marked as company (`is_company`), instead of individual, having a parent_id.  See the computation of `commercial_parent_id` as reference:

https://github.com/odoo/odoo/blob/3c45f08825b1a0de35f2027f7cb53977f2c11f08/odoo/addons/base/models/res_partner.py#L276

A possible domain would be:
```
'|',
('parent_id', '=', False),
('is_company', '=', True),
```

File: account_financial_report/wizard/general_ledger_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/general_ledger_wizard.py]: `partner_ids |= set(corp_partners.mapped('commercial_partner_id.id'))`

But it's better to work full method with recordsets:
```python
partners = self.env['res.partner'].browse(context['active_ids'])
corp_partners = partners.filtered('parent_id')
return partners - corp_partners + corp_partners.mapped('commercial_partner_id')
```

File: account_financial_report/wizard/general_ledger_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/general_ledger_wizard.py]: `fields.Date.to_string`

File: account_financial_report/wizard/general_ledger_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/general_ledger_wizard.py]: The same about the domain

File: account_financial_report/wizard/open_items_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/open_items_wizard.py]: The same about how to get partners through recordsets

File: account_financial_report/wizard/open_items_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/open_items_wizard.py]: The same about domain

File: account_financial_report/wizard/trial_balance_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/trial_balance_wizard.py]: `fields.Date.to_string`

File: account_financial_report/wizard/trial_balance_wizard.py
Comment (pedrobaeza) [inline account_financial_report/wizard/trial_balance_wizard.py]: The same about the domain. As this domain is repeated through a lot of places, I think it should be extracted to a method in the abstract model or similar.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: account_financial_report/menuitems.xml
Comment (eLBati) [reply account_financial_report/menuitems.xml]: They are. Fixing

File: account_financial_report/report/abstract_report.py
Comment (eLBati) [reply account_financial_report/report/abstract_report.py]: How is this affected by SQL injection? `self._table` doesn't come from user input

Comment (pedrobaeza) [reply account_financial_report/report/abstract_report.py]: Someone can manipulate that python variable. In other place you have replace it passing that parameter in the list. I suggested to put it with `AsIs` method for avoiding the need to disable the linter.

Comment (eLBati) [reply account_financial_report/report/abstract_report.py]: Like this
```
        query = AsIs(
            "DELETE FROM " + self._table +
            " WHERE COALESCE("
            "write_date, create_date, (now() at time zone 'UTC'))"
            "::timestamp < ((now() at time zone 'UTC') - interval %s)"
        )
```
?

> Someone can manipulate that python variable

I wonder how

> for avoiding the need to disable the linter

It is not disabled here. Am I missing something?

Comment (pedrobaeza) [reply account_financial_report/report/abstract_report.py]: No, I mean:
```python
query = "DELETE FROM %s..."
cr.execute(query, (AsIs(self._table), ...))
```

File: account_financial_report/report/open_items.py
Comment (eLBati) [reply account_financial_report/report/open_items.py]: To avoid disabling pylint, I think here we should remove variables like `where_condition_partner_by_account` and write the string directly at the end of `query_computer_partner_residual_cumul` and `query_compute_partners_due_cumul`, right?

Comment (pedrobaeza) [reply account_financial_report/report/open_items.py]: Well, I haven't seen full code right now, but you know the basis. Do what you consider best :smiley:

**Comment by eLBati** [comment]
@pedrobaeza please have a look at fdaa4848d44bc40523c7d7a4e970ead495b6107b
Thanks!

File: account_financial_report/tests/abstract_test.py
Comment (eLBati) [reply account_financial_report/tests/abstract_test.py]: I don't understand, tests are green. How could another chart be present?

Comment (pedrobaeza) [reply account_financial_report/tests/abstract_test.py]: When you are executing your tests in a customer environment CI (aka more or less integration tests) that has for example `l10n_es` installed.

Comment (eLBati) [reply account_financial_report/tests/abstract_test.py]: I installed `l10n_es` in a demo DB, executed tests of `account_financial_report` and no error was raised

File: account_financial_report/report/abstract_report_xlsx.py
Comment (eLBati) [reply account_financial_report/report/abstract_report_xlsx.py]: Done

File: account_financial_report/report/abstract_report.py
Comment (eLBati) [reply account_financial_report/report/abstract_report.py]: Done

File: account_financial_report/wizard/general_ledger_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/general_ledger_wizard.py]: Done

File: account_financial_report/wizard/open_items_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/open_items_wizard.py]: Done

File: account_financial_report/menuitems.xml
Comment (eLBati) [reply account_financial_report/menuitems.xml]: Done

File: account_financial_report/tests/test_general_ledger.py
Comment (eLBati) [reply account_financial_report/tests/test_general_ledger.py]: Done

File: account_financial_report/wizard/general_ledger_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/general_ledger_wizard.py]: Done

File: account_financial_report/wizard/trial_balance_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/trial_balance_wizard.py]: Done

File: account_financial_report/report/abstract_report.py
Comment (eLBati) [reply account_financial_report/report/abstract_report.py]: Done

File: account_financial_report/report/open_items.py
Comment (eLBati) [reply account_financial_report/report/open_items.py]: Done

File: account_financial_report/wizard/aged_partner_balance_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/aged_partner_balance_wizard.py]: Done

File: account_financial_report/wizard/general_ledger_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/general_ledger_wizard.py]: Done

File: account_financial_report/wizard/general_ledger_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/general_ledger_wizard.py]: Done

File: account_financial_report/wizard/open_items_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/open_items_wizard.py]: Done

File: account_financial_report/wizard/trial_balance_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/trial_balance_wizard.py]: Done

File: account_financial_report/wizard/trial_balance_wizard.py
Comment (eLBati) [reply account_financial_report/wizard/trial_balance_wizard.py]: Done

File: account_financial_report/tests/abstract_test.py
Comment (eLBati) [reply account_financial_report/tests/abstract_test.py:26]: If I remove it, I got
```
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: ` Check if report HTML is correctly generated 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: Traceback (most recent call last): 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `   File "/home/elbati/workspace/odoo/instances/odoo12/parts/account-financial-reporting/account_financial_report/tests/abstract_test.py", line 185, in setUp 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `     self._add_chart_of_accounts() 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `   File "/home/elbati/workspace/odoo/instances/odoo12/parts/account-financial-reporting/account_financial_report/tests/abstract_test.py", line 96, in _add_chart_of_accounts 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `     self.chart.try_loading_for_current_company() 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `   File "<decorator-gen-184>", line 2, in try_loading_for_current_company 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `   File "/home/elbati/workspace/odoo/instances/odoo12/parts/odoo/odoo/api.py", line 372, in loop 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `     result = [method(rec, *args, **kwargs) for rec in self] 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `   File "/home/elbati/workspace/odoo/instances/odoo12/parts/odoo/odoo/api.py", line 372, in <listcomp> 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `     result = [method(rec, *args, **kwargs) for rec in self] 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `   File "/home/elbati/workspace/odoo/instances/odoo12/parts/odoo/addons/account/models/chart_template.py", line 178, in try_loading_for_current_company 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `     self.load_for_current_company(15.0, 15.0) 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `   File "/home/elbati/workspace/odoo/instances/odoo12/parts/odoo/addons/account/models/chart_template.py", line 252, in load_for_current_company 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `     company.transfer_account_id = self.env['account.account'].search([('code', '=like', self.transfer_account_code_prefix + '%')])[0] 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `   File "/home/elbati/workspace/odoo/instances/odoo12/parts/odoo/odoo/models.py", line 5121, in __getitem__ 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: `     return self._browse((self._ids[key],), self.env) 
2019-06-13 14:19:06,488 2518 ERROR demo_12 odoo.addons.account_financial_report.tests.test_aged_partner_balance: ` IndexError: tuple index out of range
```

Comment (pedrobaeza) [reply account_financial_report/tests/abstract_test.py:26]: It should be created in https://github.com/odoo/odoo/blob/a526ff3f70c472fcd7647e1b7b75dcfa9d539209/addons/account/models/chart_template.py#L130

I'll investigate locally and change it at the same time as I squash some commits before merging.

Comment (pedrobaeza) [reply account_financial_report/tests/abstract_test.py:26]: OK, I'm not able to find the reason, but on l10n_es works without this. Let's move on as no real problem to let that as they are tests.

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Thanks all for your work!

**Pull request merged**
